extensions/statistics/js/src/admin/components/StatisticsWidgetDateSelectionModal.tsx

```diff
@@ -0,0 +1,149 @@
+import app from 'flarum/admin/app';
+import ItemList from 'flarum/common/utils/ItemList';
+import generateElementId from 'flarum/admin/utils/generateElementId';
+import Modal, { IInternalModalAttrs } from 'flarum/common/components/Modal';
+
+import Mithril from 'mithril';
+import Button from 'flarum/common/components/Button';
+
+import dayjs from 'dayjs';
+import dayjsUtc from 'dayjs/plugin/utc';
+
+dayjs.extend(dayjsUtc);
+
+export interface IDateSelection {
+  /**
+   * Timestamp (seconds, not ms) for start date
+   */
+  start: number;
+  /**
+   * Timestamp (seconds, not ms) for end date
+   */
+  end: number;
+}
+
+export interface IStatisticsWidgetDateSelectionModalAttrs extends IInternalModalAttrs {
+  onModalSubmit: (dates: IDateSelection) => void;
+  value?: IDateSelection;
+}
+
+interface IStatisticsWidgetDateSelectionModalState {
+  inputs: {
+    startDateVal: string;
+    endDateVal: string;
+  };
+  ids: {
+    startDate: string;
+    endDate: string;
+  };
+}
+
+export default class StatisticsWidgetDateSelectionModal extends Modal<IStatisticsWidgetDateSelectionModalAttrs> {
+  /* @ts-expect-error core typings don't allow us to set the type of the state attr :( */
```

Comment:
noticed this too, I guess like we have to always type a default state/attribute type in any core components to not hinder extensibility going forward?

Reply:
Yeah, we've done it in some components, but not all.

Just a simple `class MyComponent<CustomAttrs extends BaseAttrs = BaseAttrs, CustomState extends BaseState> extends BaseComponent<CustomAttrs, CustomState>` should work.